<a href="https://colab.research.google.com/github/DavisKerr/cse380-notebooks/blob/master/10_3_About_Patterns_and_Probabilities.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# About Patterns and Probabilities
## Class Directed Learning
### Due: Tuesday, 9 March 2021, 11:59 pm

## TODO Explore and Wonder

Regarding spanning trees of ladder graphs:

What is the probability ($\lim_{n \rightarrow \infty} P(n)$) that a randomly-selected spanning tree of an $n$-rung ladder graph contains the bottom rung?

| n | P(n) |
|---|------|
| 1 |   1   |
| 2 |   .75   |
| 3 |   .733   |
| 4 |    .732  |
| 5 |    .732  |


In [1]:
from math import factorial as fact
from math import sqrt, ceil
def nCk(n, k):
  return (fact(n)) / ((fact(n-k) * fact(k)))

def NST2(n):
  return ((((2 + sqrt(3)) ** n) - (2 - sqrt(3)) ** n) / (2 * sqrt(3)))

In [5]:
def P(n):
  return (NST2(n) - NST2(n-1)) / (NST2(n))

In [8]:
for n in range(1,100):
  print(P(n))

1.0
0.75
0.7333333333333333
0.7321428571428571
0.7320574162679426
0.732051282051282
0.7320508416351769
0.7320508100147275
0.7320508077444814
0.7320508075814851
0.7320508075697825
0.7320508075689423
0.7320508075688819
0.7320508075688776
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.7320508075688774
0.7320508075688772
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.7320508075688774
0.7320508075688773
0.7320508075688773
0.7320508075688772
0.7320508075688773
0.7320508075688773
0.7320508075688772
0.7320508075688772
0.7320508075688773
0.7320508075688772
0.7320508075688773
0.7320508075688772
0.7320508075688773
0.7320508075688772
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.7320508075688773
0.73

There is a noticable limit at 73.2%. This means there is a 73.2% chance of randomly selecting a spanning tree from an n-tier ladder graph that uses the bottom rung.

$ lim_{5 \rightarrow  \inf}  1 - \frac{\frac{(2+\sqrt{3})^{n-1} - (2-\sqrt{3})^{n-1}}{2\sqrt{3}}}{\frac{(2+\sqrt{3})^{n} - (2-\sqrt{3})^{n}}{2\sqrt{3}}}$

### Recurrence Relations

Let $f(n) =$ NSTIBR$(n)$:

$f(n) = 4f(n-1) - f(n-2)$ for $n > 1$;

$f(0) = 1$,

$f(1) = 1$.

Let $g(n) =$ NST$(n)$:

$g(n) = 4g(n-1) - g(n-2)$ for $n > 1$;

$g(0) = 0$,

$g(1) = 1$.

#### TODO Check Recurrences

In [18]:
def g(n):
  if n < 0:
    return "N/A"
  if n == 0:
    return 0
  elif n == 1:
    return 1
  return 4 * g(n - 1) - g(n-2)

def f(n):
  if n < 0:
    return "N/A"
  if n == 0 or n == 1:
    return 1
  return 4 * f(n-1) - f(n-2)

In [19]:
for n in range(0, 6):
  print(f(n))

1
1
3
11
41
153


In [20]:
for n in range(0, 6):
  print(f(n - 1))

N/A
1
1
3
11
41


In [21]:
for n in range(0, 6):
  print(4 * f(n - 1))

N/AN/AN/AN/A
4
4
12
44
164


In [22]:
for n in range(0, 6):
  print(f(n - 2))

N/A
N/A
1
1
3
11


In [23]:
for n in range(0, 6):
  print(g(n))

0
1
4
15
56
209


In [ ]:
for n in range(0, 6):
  print(g(n))

In [24]:
for n in range(0, 6):
  print(4 * g(n-1))

N/AN/AN/AN/A
0
4
16
60
224


In [25]:
for n in range(0, 6):
  print(f(n-2))

N/A
N/A
1
1
3
11


Check the calculations in this table (maybe write recursive functions) to verify they are accurate.

Do they match what you found yesterday in your DPC?

| n | f(n) | f(n-1) | 4f(n-1)  | f(n - 2) | diff |
|---|-----:|-------:|---------:|---------:|-----:|
| 0 |    1 |    N/A |      N/A |      N/A |  N/A |
| 1 |    1 |      1 |        4 |      N/A |  N/A |
| 2 |    3 |      1 |        4 |        1 |    3 |
| 3 |   11 |      3 |       12 |        1 |   11 |
| 4 |   41 |     11 |       44 |        3 |   41 |
| 5 |  153 |     41 |      164 |       11 |  153 |

| n | g(n) | g(n-1) | 4g(n-1)  | g(n - 2) | diff |
|---|-----:|-------:|---------:|---------:|-----:|
| 0 |    0 |    N/A |      N/A |      N/A |  N/A |
| 1 |    1 |      0 |        0 |      N/A |  N/A |
| 2 |    4 |      1 |        4 |        0 |    4 |
| 3 |   15 |      4 |       16 |        1 |   15 |
| 4 |   56 |     15 |       60 |        4 |   56 |
| 5 |  209 |     56 |      224 |       15 |  209 |

### TODO Find closed-form formulas

Can you find closed-form formulas for $f(n)$ and $g(n)$?

A closed-form formula expressing these functions in terms of operations on $n$, **without** referring to previous calculated values of the functions.

$$g(n) = \frac{(2+\sqrt{3})^{n} - (2-\sqrt{3})^{n}}{2\sqrt{3}}$$
$$f(n) = g(n) - g(n-1) $$

#### Hint:

http://www.ist.tugraz.at/aichholzer/teaching/eca/spanning_trees_in_ladders.pdf

#### Possibly Illuminating Calculations

Remember continued fractions?

In [ ]:
from fractions import Fraction as frac

def contfrac2frac(seq):
    """Convert the simple continued fraction in `seq`
       into a fraction with numerator num and denominator den.
    """
    num, den = 1, 0
    for u in reversed(seq):
        num, den = den + num * u, num
    return frac(num, den)

def frac2contfrac(f):
    """Build the simple continued fraction expansion of fraction f.
    """
    seq = []
    frac2contfrac_rec(f, seq)
    return seq

def frac2contfrac_rec(f, seq):
    n = f.numerator
    d = f.denominator
    if d != 0:
        seq.append(n // d)
        if n % d != 0:
            frac2contfrac_rec(frac(d, n % d), seq)

def eval_frac(f):
    """Evaluate the fraction f as a float.
    """
    return f.numerator / f.denominator

In [ ]:
from math import sqrt

value = sqrt(3) - 1
value_as_cf = frac2contfrac(frac.from_float(value))[:21]
cf_to_value = contfrac2frac(value_as_cf)
print(value, value_as_cf, cf_to_value)
print(eval_frac(cf_to_value))

In [ ]:
from pprint import pprint

pprint(list(map(lambda n: contfrac2frac(value_as_cf[:n]), range(20, 5, -1))))